pkg/querier/block_store.go

```diff
@@ -83,41 +82,44 @@ func NewUserStore(cfg tsdb.Config, bucketClient objstore.Bucket, logLevel loggin
 		registerer.MustRegister(u.syncTimes, u.tsdbMetrics)
 	}
 
-	serv := grpc.NewServer()
-	storepb.RegisterStoreServer(serv, u)
+	u.Service = services.NewBasicService(u.starting, u.syncStoresLoop, u.stopping)
+	return u, nil
+}
+
+func (u *UserStore) starting(ctx context.Context) error {
+	u.serv = grpc.NewServer()
+	storepb.RegisterStoreServer(u.serv, u)
 	l, err := net.Listen("tcp", "")
 	if err != nil {
-		return nil, err
+		return err
 	}
-	go serv.Serve(l)
+	go func() {
+		err := u.serv.Serve(l)
+		if err != nil {
+			level.Error(u.logger).Log("msg", "block store grpc server failed", "err", err)
+		}
+	}()
 
 	cc, err := grpc.Dial(l.Addr().String(), grpc.WithInsecure())
 	if err != nil {
-		return nil, err
+		return err
 	}
 
 	u.client = storepb.NewStoreClient(cc)
 
-	// If the sync is disabled we never sync blocks, which means the bucket store
-	// will be empty and no series will be returned once queried.
 	if u.cfg.BucketStore.SyncInterval > 0 {
 		// Run an initial blocks sync, required in order to be able to serve queries.
-		if err := u.initialSync(workersCtx); err != nil {
-			return nil, err
+		if err := u.initialSync(ctx); err != nil {
+			return err
 		}
-
-		// Periodically sync the blocks.
-		u.workers.Add(1)
-		go u.syncStoresLoop(workersCtx)
 	}
 
-	return u, nil
+	return nil
 }
 
-// Stop the blocks sync and waits until done.
-func (u *UserStore) Stop() {
-	u.workersCancel()
-	u.workers.Wait()
```

Comment:
There was only one worker, `syncStoresLoop`.

`stopping` method is now only called after `syncStoresLoop` exits (because that is "running function" of this service), so yes, wait logic is preserved.